File: client-py/queen/http/load_balancer.py
"""
Load balancer for distributing requests across multiple servers
Supports round-robin, session, and affinity-based routing with virtual nodes
"""

import time
from typing import Dict, List, Optional, Any


class LoadBalancer:
    """Load balancer with round-robin, session, and affinity strategies"""

    def __init__(
        self,
        urls: List[str],
        strategy: str = "round-robin",
        *,
        affinity_hash_ring: int = 128,
        health_retry_after_millis: int = 5000,
    ):
        """
        Initialize load balancer

        Args:
            urls: List of server URLs
            strategy: Load balancing strategy ('round-robin', 'session', 'affinity')
            affinity_hash_ring: Number of virtual nodes per server for affinity
            health_retry_after_millis: Retry unhealthy backends after N milliseconds
        """
        self._urls = [url.rstrip("/") for url in urls]  # Remove trailing slashes
        self._strategy = strategy
        self._current_index = 0
        self._session_map: Dict[str, int] = {}
        self._session_id = f"session_{int(time.time())}_{id(self)}"
        self._affinity_hash_ring = affinity_hash_ring
        self._health_retry_after_millis = health_retry_after_millis

        # Health status: url -> {'healthy': bool, 'failures': int, 'last_failure': float}
        self._health_status: Dict[str, Dict[str, Any]] = {
            url: {"healthy": True, "failures": 0, "last_failure": None} for url in self._urls
        }

        # Virtual nodes for affinity strategy
        self._virtual_nodes: Optional[List[Dict[str, Any]]] = None

        if strategy == "affinity":
            self._build_virtual_node_ring()

    def _build_virtual_node_ring(self) -> None:
        """Build virtual node ring for consistent hashing"""
        self._virtual_nodes = []

        for url in self._urls:
            for i in range(self._affinity_hash_ring):
                # Create virtual node identifier
                vnode_key = f"{url}#vnode{i}"
                hash_value = self._hash_string(vnode_key)

                self._virtual_nodes.append({"hash": hash_value, "real_server": url})

        # Sort by hash value to create the ring
        self._virtual_nodes.sort(key=lambda x: x["hash"])

    def _rebuild_virtual_node_ring(self, urls: Optional[List[str]] = None) -> None:
        """
        Rebuild virtual node ring with specified URLs

        Args:
            urls: URLs to include in ring (if None, uses all healthy URLs)
        """
        urls_to_include = (
            urls
            if urls is not None
            else [url for url in self._urls if self._health_status[url]["healthy"]]
        )

        self._virtual_nodes = []

        for url in urls_to_include:
            for i in range(self._affinity_hash_ring):
                vnode_key = f"{url}#vnode{i}"
                hash_value = self._hash_string(vnode_key)

                self._virtual_nodes.append({"hash": hash_value, "real_server": url})

        self._virtual_nodes.sort(key=lambda x: x["hash"])

    def get_next_url(self, session_key: Optional[str] = None) -> str:
        """
        Get next URL based on strategy

        Args:
            session_key: Optional session key for affinity or session routing

        Returns:
            Next URL to use
        """
        key = session_key or self._session_id

        # Get healthy URLs + unhealthy URLs that are ready for retry
        now = time.time()
        need_ring_rebuild = False

        healthy_urls = []
        for url in self._urls:
            status = self._health_status[url]
            if status["healthy"]:
                healthy_urls.append(url)
            else:
                # Allow retry of unhealthy backends after configured interval
                last_failure = status["last_failure"]
                if last_failure and (now - last_failure) >= (
                    self._health_retry_after_millis / 1000.0
                ):
                    need_ring_rebuild = True  # Need to rebuild ring to include this server
                    healthy_urls.append(url)

        # Rebuild vnode ring if we have servers eligible for retry
        if need_ring_rebuild and self._virtual_nodes is not None:
            self._rebuild_virtual_node_ring(healthy_urls)

        if not healthy_urls:
            # All backends unhealthy, try any backend as fallback
            return self._urls[self._current_index % len(self._urls)]

        if self._strategy == "affinity":
            # Virtual node consistent hashing
            return self._get_affinity_url(key, healthy_urls)

        if self._strategy == "session":
            # Session affinity: stick to the same server per session
            if key not in self._session_map:
                assigned_index = self._current_index
                self._session_map[key] = assigned_index
                self._current_index = (self._current_index + 1) % len(healthy_urls)

            assigned_url = self._urls[self._session_map[key]]
            # If assigned URL is unhealthy, reassign
            if assigned_url not in healthy_urls:
                new_index = self._current_index % len(healthy_urls)
                self._session_map[key] = new_index
                self._current_index = (self._current_index + 1) % len(healthy_urls)
                return healthy_urls[new_index]

            return assigned_url

        # Round robin: cycle through healthy URLs only
        url = healthy_urls[self._current_index % len(healthy_urls)]
        self._current_index = (self._current_index + 1) % len(healthy_urls)
        return url

    def _get_affinity_url(self, key: str, healthy_urls: List[str]) -> str:
        """
        Affinity-based routing using consistent hashing with virtual nodes

        Args:
            key: Affinity key
            healthy_urls: List of healthy URLs

        Returns:
            URL to use
        """
        if not key or not self._virtual_nodes or not self._virtual_nodes:
            # No key or no vnodes, fall back to round-robin
            url = healthy_urls[self._current_index % len(healthy_urls)]
            self._current_index = (self._current_index + 1) % len(healthy_urls)
            return url

        # Hash the consumer group key
        key_hash = self._hash_string(key)

        # Binary search to find the first vnode >= keyHash (clockwise on ring)
        left = 0
        right = len(self._virtual_nodes) - 1
        result = 0

        while left <= right:
            mid = (left + right) // 2
            if self._virtual_nodes[mid]["hash"] >= key_hash:
                result = mid
                right = mid - 1
            else:
                left = mid + 1

        # Walk forward on the ring to find first healthy server
        start_idx = result
        for i in range(len(self._virtual_nodes)):
            idx = (start_idx + i) % len(self._virtual_nodes)
            vnode = self._virtual_nodes[idx]

            if vnode["real_server"] in healthy_urls:
                return vnode["real_server"]

        # Fallback (should never happen if healthy_urls is not empty)
        return healthy_urls[0]

    def _hash_string(self, s: str) -> int:
        """
        FNV-1a hash function for consistent hashing

        Args:
            s: String to hash

        Returns:
            Unsigned 32-bit integer hash
        """
        hash_value = 2166136261  # FNV offset basis
        for char in s:
            hash_value ^= ord(char)
            hash_value += (
                (hash_value << 1)
                + (hash_value << 4)
                + (hash_value << 7)
                + (hash_value << 8)
                + (hash_value << 24)
            )
            # Keep it in 32-bit range
            hash_value &= 0xFFFFFFFF
        return hash_value

    def mark_unhealthy(self, url: str) -> None:
        """
        Mark a backend as unhealthy after failure

        Args:
            url: URL to mark unhealthy
        """
        status = self._health_status.get(url)
        if status:
            status["healthy"] = False
            status["failures"] += 1
            status["last_failure"] = time.time()

            # Rebuild vnode ring to exclude unhealthy server
            if self._virtual_nodes is not None:
                self._rebuild_virtual_node_ring()

    def mark_healthy(self, url: str) -> None:
        """
        Mark a backend as healthy after success

        Args:
            url: URL to mark healthy
        """
        status = self._health_status.get(url)
        if not status:
            return

        was_unhealthy = not status["healthy"]

        status["healthy"] = True
        status["failures"] = 0
        status["last_failure"] = None

        # Rebuild vnode ring to include recovered server
        if self._virtual_nodes is not None and was_unhealthy:
            self._rebuild_virtual_node_ring()

    def get_health_status(self) -> Dict[str, Dict[str, Any]]:
        """Get health status of all backends"""
        return dict(self._health_status)

    def get_all_urls(self) -> List[str]:
        """Get all URLs"""
        return list(self._urls)

    def get_strategy(self) -> str:
        """Get load balancing strategy"""
        return self._strategy

    def get_virtual_node_count(self) -> int:
        """Get number of virtual nodes in the ring"""
        return len(self._virtual_nodes) if self._virtual_nodes else 0

    def reset(self) -> None:
        """Reset load balancer state"""
        self._current_index = 0
        self._session_map.clear()

File: clients/client-js/client-v2/http/LoadBalancer.js
/**
 * Load balancer for distributing requests across multiple servers
 * Supports round-robin, session, and affinity-based routing with virtual nodes
 */

export class LoadBalancer {
  #urls
  #strategy
  #currentIndex = 0
  #sessionMap = new Map()
  #sessionId
  #healthStatus
  #virtualNodes = null
  #affinityHashRing
  #healthRetryAfterMillis

  constructor(urls, strategy = 'round-robin', options = {}) {
    this.#urls = urls.map(url => url.replace(/\/$/, '')) // Remove trailing slashes
    this.#strategy = strategy
    this.#sessionId = `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`
    
    // Virtual node replicas for affinity strategy (default: 150)
    this.#affinityHashRing = options.affinityHashRing || 150
    
    // Health retry interval (default: 30 seconds)
    this.#healthRetryAfterMillis = options.healthRetryAfterMillis || 30000
    
    // Initialize health status with last failure time
    this.#healthStatus = new Map(this.#urls.map(url => [url, { 
      healthy: true, 
      failures: 0,
      lastFailure: null
    }]))
    
    // Build virtual node ring for affinity strategy
    if (strategy === 'affinity') {
      this.#buildVirtualNodeRing()
    }
  }

  /**
   * Build virtual node ring for consistent hashing
   * Each real server gets multiple virtual positions in the hash space
   */
  #buildVirtualNodeRing() {
    this.#virtualNodes = []
    
    for (const url of this.#urls) {
      for (let i = 0; i < this.#affinityHashRing; i++) {
        // Create virtual node identifier
        const vnodeKey = `${url}#vnode${i}`
        const hash = this.#hashString(vnodeKey)
        
        this.#virtualNodes.push({
          hash,
          realServer: url
        })
      }
    }
    
    // Sort by hash value to create the ring
    this.#virtualNodes.sort((a, b) => a.hash - b.hash)
  }

  /**
   * Rebuild virtual node ring with specified URLs
   * @param {Array<string>} urls - URLs to include in the ring (if not provided, uses all healthy URLs)
   */
  #rebuildVirtualNodeRing(urls = null) {
    const urlsToInclude = urls || this.#urls.filter(url => this.#healthStatus.get(url)?.healthy !== false)
    
    this.#virtualNodes = []
    
    for (const url of urlsToInclude) {
      for (let i = 0; i < this.#affinityHashRing; i++) {
        const vnodeKey = `${url}#vnode${i}`
        const hash = this.#hashString(vnodeKey)
        
        this.#virtualNodes.push({
          hash,
          realServer: url
        })
      }
    }
    
    this.#virtualNodes.sort((a, b) => a.hash - b.hash)
  }

  getNextUrl(sessionKey = null) {
    const key = sessionKey || this.#sessionId
    
    // Get healthy URLs + unhealthy URLs that are ready for retry
    const now = Date.now()
    let needRingRebuild = false
    
    const healthyUrls = this.#urls.filter(url => {
      const status = this.#healthStatus.get(url)
      if (status.healthy) return true
      
      // Allow retry of unhealthy backends after configured interval
      if (status.lastFailure && (now - status.lastFailure) >= this.#healthRetryAfterMillis) {
        needRingRebuild = true  // Need to rebuild ring to include this server's vnodes
        return true
      }
      
      return false
    })
    
    // Rebuild vnode ring if we have servers eligible for retry
    // Pass healthyUrls so it includes servers ready for retry
    if (needRingRebuild && this.#virtualNodes) {
      this.#rebuildVirtualNodeRing(healthyUrls)
    }
    
    if (healthyUrls.length === 0) {
      // All backends unhealthy, try any backend as fallback
      return this.#urls[this.#currentIndex % this.#urls.length]
    }

    if (this.#strategy === 'affinity') {
      // Virtual node consistent hashing
      return this.#getAffinityUrl(key, healthyUrls)
    }
    
    if (this.#strategy === 'session') {
      // Session affinity: stick to the same server per session
      if (!this.#sessionMap.has(key)) {
        const assignedIndex = this.#currentIndex
        this.#sessionMap.set(key, assignedIndex)
        this.#currentIndex = (this.#currentIndex + 1) % healthyUrls.length
      }
      
      const assignedUrl = this.#urls[this.#sessionMap.get(key)]
      // If assigned URL is unhealthy, reassign
      if (!healthyUrls.includes(assignedUrl)) {
        const newIndex = this.#currentIndex % healthyUrls.length
        this.#sessionMap.set(key, newIndex)
        this.#currentIndex = (this.#currentIndex + 1) % healthyUrls.length
        return healthyUrls[newIndex]
      }
      
      return assignedUrl
    }

    // Round robin: cycle through healthy URLs only
    const url = healthyUrls[this.#currentIndex % healthyUrls.length]
    this.#currentIndex = (this.#currentIndex + 1) % healthyUrls.length
    return url
  }

  /**
   * Affinity-based routing using consistent hashing with virtual nodes
   * Same key always routes to same backend (unless unhealthy)
   */
  #getAffinityUrl(key, healthyUrls) {
    if (!key || !this.#virtualNodes || this.#virtualNodes.length === 0) {
      // No key or no vnodes, fall back to round-robin
      return healthyUrls[this.#currentIndex++ % healthyUrls.length]
    }
    
    // Hash the consumer group key
    const keyHash = this.#hashString(key)
    
    // Binary search to find the first vnode >= keyHash (clockwise on ring)
    let left = 0
    let right = this.#virtualNodes.length - 1
    let result = 0
    
    while (left <= right) {
      const mid = Math.floor((left + right) / 2)
      if (this.#virtualNodes[mid].hash >= keyHash) {
        result = mid
        right = mid - 1
      } else {
        left = mid + 1
      }
    }
    
    // Walk forward on the ring to find first healthy server
    const startIdx = result
    for (let i = 0; i < this.#virtualNodes.length; i++) {
      const idx = (startIdx + i) % this.#virtualNodes.length
      const vnode = this.#virtualNodes[idx]
      
      if (healthyUrls.includes(vnode.realServer)) {
        return vnode.realServer
      }
    }
    
    // Fallback (should never happen if healthyUrls is not empty)
    return healthyUrls[0]
  }

  /**
   * FNV-1a hash function for consistent hashing
   */
  #hashString(str) {
    let hash = 2166136261 // FNV offset basis
    for (let i = 0; i < str.length; i++) {
      hash ^= str.charCodeAt(i)
      hash += (hash << 1) + (hash << 4) + (hash << 7) + (hash << 8) + (hash << 24)
    }
    return hash >>> 0 // Convert to unsigned 32-bit integer
  }

  /**
   * Mark a backend as unhealthy after failure
   */
  markUnhealthy(url) {
    const status = this.#healthStatus.get(url)
    if (status) {
      status.healthy = false
      status.failures++
      status.lastFailure = Date.now()
      
      const retryInSeconds = (this.#healthRetryAfterMillis / 1000).toFixed(0)
      
      // Rebuild vnode ring to exclude unhealthy server
      if (this.#virtualNodes) {
        this.#rebuildVirtualNodeRing()
      }
    }
  }

  /**
   * Mark a backend as healthy after success
   */
  markHealthy(url) {
    const status = this.#healthStatus.get(url)
    if (!status) return
    
    const wasUnhealthy = status.healthy === false
    
    status.healthy = true
    status.failures = 0
    status.lastFailure = null
    
    // Rebuild vnode ring to include recovered server
    if (this.#virtualNodes && wasUnhealthy) {
      this.#rebuildVirtualNodeRing()
    }
  }

  /**
   * Get health status of all backends
   */
  getHealthStatus() {
    return new Map(this.#healthStatus)
  }

  getAllUrls() {
    return [...this.#urls]
  }

  getStrategy() {
    return this.#strategy
  }
  
  /**
   * Get number of virtual nodes in the ring
   */
  getVirtualNodeCount() {
    return this.#virtualNodes ? this.#virtualNodes.length : 0
  }

  reset() {
    this.#currentIndex = 0
    this.#sessionMap.clear()
  }
}

